README.md
# DBERlibR
### An R Package for Automated Assessment Data Analysis

DBERlibR is an R package developed by Helikar Lab at the University of Nebraska-Lincoln for automated assessment data analyses. Some of the most frequently used statistical techniques are developed into functions to clean the data, merge/bind multiple data sets (as necessary), check assumption(s) for a specific statistical technique (as necessary), run the main assessment data analysis, and print the outputs in R console. The outputs contain a sample interpretation of the results for the convenience of users. Users need to prepare the data file as instructed.

# Functions

### item_analysis(score_csv_data, m_cutoff = 0.15)
This function automatically reads and cleans the data (e.g., converting missing values to “0”), calculates difficulty and discriminant scores, and prints the results in the R console and Plots panel. 

### paired_samples(pre_csv_data, post_csv_data, m_cutoff = 0.15)
This function automatically reads and cleans the data sets (e.g., converting missing values to "0), merges pre-post data sets, runs the (parametric) paired samples T-test and (nonparametric) Wilcoxon signed-rank test, and then prints outputs in the R console and Plots panel to help users examine the difference between pre-post scores. 

### independent_samples(treat_csv_data, ctrl_csv_data, m_cutoff = 0.15)
This function automatically reads and cleans the data sets (e.g., converting missing values to "0), binds treatment-control group data sets, runs the independent samples t-test (parametric) and Mann–Whitney U test (nonparametric), and then prints the results in the R console and Plots panel to help users examine the difference between the groups. 

### one_way_ancova( treat_pre_csv_data, treat_post_csv_data, ctrl_pre_csv_data, ctrl_post_csv_data, m_cutoff = 0.15)
This function can be used to analyze the difference between two groups (e.g., intervention vs. control group) with a covariate (e.g., pre-test scores) controlled. The function automatically merges pre-post data sets, binds treatment-control data sets, runs scripts to check assumptions of one-way ANCOVA, runs the main one-way ANCOVA and post hoc analyses, and then displays all outputs in the R console and Plots panel for users all at once. 

### one_way_repeated_anova(treat_pre_csv_data, treat_post_csv_data, treat_post2_csv_data, m_cutoff = 0.15)
This function can be used when you collect data from the same students repeatedly at three different time points (e.g., pre-test, post-test, and post2-test) and you want to examine the significance of the changes over time. The function automatically merges pre, post, and post2 data sets, runs the one-way repeated measures ANOVA with assumptions check, and then displays the outputs in the R console and Plots panel for users all at once.

### demo_group_diff(score_csv_data, group_csv_data, m_cutoff = 0.15, group_name)
This function automatically combines demographic variables to a data set, runs the analysis of variance (ANOVA) with assumptions check to examine demographic sub-group differences, and then displays the outputs in the R console and Plots panel for users all at once.


NEWS.md
## version 0.1.0

New release - features 
- item_analysis(score_csv_data, m_cutoff = 0.15)
- paired_samples(pre_csv_data, post_csv_data, m_cutoff = 0.15)
- independent_samples(treat_csv_data, ctrl_csv_data, m_cutoff = 0.15)
- one_way_ancova(treat_pre_csv_data, treat_post_csv_data, ctrl_pre_csv_data, ctrl_post_csv_data, m_cutoff = 0.15)
- one_way_repeated_anova(treat_pre_csv_data, treat_post_csv_data, treat_post2_csv_data, m_cutoff = 0.15)
- demo_group_diff(score_csv_data, group_csv_data, m_cutoff = 0.15, group_name)

### Automated analysis of assessment data for discipline-based education research.

- Users need to set the folder where their data to import resides as the working directory if they include only data file name(s) in the function.
- Alternatively, users can provide a path to their data in the function.

### NEWS.md setup

- added NEWS.md creation with newsmd


## version 0.1.1

"+ file LICENSE" deleted from the Description: License. 

## version 0.1.2

- Wrote package names in single quotes in the title and description field
- Added a reference on the methods used in the package to the description field 
- Added \value to .Rd files
- Deleted options(warn=-1)
- Used message() to write information messages
- Deleted the code to install packages in the vignette

## version 0.1.3

- Fixed DOIs as <doi:10.prefix/suffix> in the description field as requested



inst/app.R/paired_samples_gui.R
# # This is a Shiny web application. You can run the application by clicking
# # the 'Run App' button above.
# #
# # Find out more about building applications with Shiny here:
# #
# #    http://shiny.rstudio.com/
# #
# #------------------------------------------------------
# # https://chat.openai.com/chat  how to build graphical user interface to take inputs from the user in an r package
# # https://deanattali.com/2015/04/21/r-package-shiny-app/
# #------------------------------------------------------

library(DBERlibR)
library(shiny)
library(shinyFiles)

ui <- fluidPage(
  navlistPanel(
    tabPanel("Item Analysis",
             fluidRow(
               column(6, fileInput("file1", "Choose Data File"),
                      actionButton("submit1", "Run Item Analysis")
               ),
               column(3, textInput("text1", "Cutoff Value"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    ),
    tabPanel("Paired Samples Data Analysis",
             fluidRow(
               column(6, fileInput("file2", "Choose Pre-test Data File"),
                      fileInput("file3", "Choose Post-test Data File"),
                      actionButton("submit2", "Run Paired Samples Data Analysis"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(3, textInput("text2", "Cutoff Value"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    ),
    tabPanel("Independent Samples Data Analysis",
             fluidRow(
               column(6, fileInput("file4", "Choose Treatment Group Data File"),
                      fileInput("file5", "Choose Control Group Data File"),
                      actionButton("submit3", "Run Independent Samples Data Analysis"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(3, textInput("text3", "Cutoff Value"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    ),
    tabPanel("One-way ANCOVA",
             fluidRow(
               column(6, fileInput("file6", "Choose Treatment-Group Pre-test Data File"),
                      fileInput("file7", "Choose Treatment Group Post-test Data File"),
                      fileInput("file8", "Choose Control-Group Pre-test Data File"),
                      fileInput("file9", "Choose Control-Group Post-test Data File"),
                      actionButton("submit4", "Run One-way ANCOVA"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(3, textInput("text4", "Cutoff Value"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    ),
    tabPanel("One-way Repeated Measures ANOVA",
             fluidRow(
               column(6, fileInput("file10", "Choose Pre-test Data File"),
                      fileInput("file11", "Choose Post-test Data File"),
                      fileInput("file12", "Choose Post2-test Data File"),
                      actionButton("submit5", "Run One-way Repeated Measures ANOVA"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(3, textInput("text5", "Cutoff Value"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    ),
    tabPanel("Analysis of Demographic Group Differences",
             fluidRow(
               column(6, fileInput("file13", "Choose Assessment Data File"),
                      fileInput("file14", "Choose Demographic Data File"),
                      actionButton("submit6", "Run Analysis of Demographic Group Differences"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(3, textInput("text6", "Cutoff Value"),
                      textInput("text7", "Group Variable Name"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               )
             )
    )
  )
)

server <- function(input, output) {
  file1 <- reactive({
    input$file1
  })
  file2 <- reactive({
    input$file2
  })
  file3 <- reactive({
    input$file3
  })
  file4 <- reactive({
    input$file4
  })
  file5 <- reactive({
    input$file5
  })
  file6 <- reactive({
    input$file6
  })
  file7 <- reactive({
    input$file7
  })
  file8 <- reactive({
    input$file8
  })
  file9 <- reactive({
    input$file9
  })
  file10 <- reactive({
    input$file10
  })
  file11 <- reactive({
    input$file11
  })
  file12 <- reactive({
    input$file12
  })
  file13 <- reactive({
    input$file13
  })
  file14 <- reactive({
    input$file14
  })

  text1 <- reactive({
    input$text1
  })
  text2 <- reactive({
    input$text2
  })
  text3 <- reactive({
    input$text3
  })
  text4 <- reactive({
    input$text4
  })
  text5 <- reactive({
    input$text5
  })
  text6 <- reactive({
    input$text6
  })
  text7 <- reactive({
    input$text7
  })

  ## Item Analysis

  observeEvent(input$submit1, {
    if (is.null(file1())) {
      return(NULL)
    }
    data1 <- file1()
    item_analysis(score_csv_data = data1$datapath,
                  m_cutoff = text1()
                  )
  })

  ## Paired Samples Data Analysis

  observeEvent(input$submit2, {

    if (is.null(file2())) {
      return(NULL)
    }
    if (is.null(file3())) {
      return(NULL)
    }

    data2 <- file2()
    data3 <- file3()
    paired_samples(pre_csv_data = data2$datapath,
                   post_csv_data = data3$datapath,
                   m_cutoff = text2()
                   )
  })

  ## Independent Samples Data Analysis

  observeEvent(input$submit3, {

    if (is.null(file4())) {
      return(NULL)
    }
    if (is.null(file5())) {
      return(NULL)
    }

    data4 <- file4()
    data5 <- file5()
    independent_samples(treat_csv_data = data4$datapath,
                        ctrl_csv_data = data5$datapath,
                        m_cutoff = text3()
                        )
  })

  ## One-way ANCOVA

  observeEvent(input$submit4, {

    if (is.null(file6())) {
      return(NULL)
    }
    if (is.null(file7())) {
      return(NULL)
    }
    if (is.null(file8())) {
      return(NULL)
    }
    if (is.null(file9())) {
      return(NULL)
    }

    data6 <- file6()
    data7 <- file7()
    data8 <- file8()
    data9 <- file9()
    one_way_ancova(treat_pre_csv_data = data6$datapath,
                   treat_post_csv_data = data7$datapath,
                   ctrl_pre_csv_data = data8$datapath,
                   ctrl_post_csv_data = data9$datapath,
                   m_cutoff = text4()
                   )
  })

  ## One-way Repeated Measures ANOVA

  observeEvent(input$submit5, {

    if (is.null(file10())) {
      return(NULL)
    }
    if (is.null(file11())) {
      return(NULL)
    }
    if (is.null(file12())) {
      return(NULL)
    }

    data10 <- file10()
    data11 <- file11()
    data12 <- file12()
    one_way_repeated_anova(treat_pre_csv_data = data10$datapath,
                           treat_post_csv_data = data11$datapath,
                           treat_post2_csv_data = data12$datapath,
                           m_cutoff = text5()
                           )
  })

  ## Demographic Group Differences

  observeEvent(input$submit6, {

    if (is.null(file13())) {
      return(NULL)
    }
    if (is.null(file14())) {
      return(NULL)
    }

    data13 <- file13()
    data14 <- file14()
    demo_group_diff(score_csv_data = data13$datapath,
                    group_csv_data = data14$datapath,
                    m_cutoff = text6(),
                    group_name = text7()
                    )
  })

}

shinyApp(ui, server)


inst/app.R/gui_app.R
# # This is a Shiny web application. You can run the application by clicking
# # the 'Run App' button above.
# #
# # Find out more about building applications with Shiny here:
# #
# #    http://shiny.rstudio.com/
# #
# #------------------------------------------------------
# # https://chat.openai.com/chat  how to build graphical user interface to take inputs from the user in an r package
# # https://deanattali.com/2015/04/21/r-package-shiny-app/
# #------------------------------------------------------

# # Take demographic variable name and create a subset of the dataframe with average score and demographic data
# message("############")
# nd <- ncol(demographic_data)
# for (i in 2:nd) {
#   message(paste0(colnames(demographic_data[i]),"=",i-1))
# }
#

library(DBERlibR)
library(shiny)
library(shinyFiles)

ui <- fluidPage(
  navlistPanel(
    tabPanel("Item Analysis",
             fluidRow(
               column(12, br(),
                      fileInput("file1", "Choose Data File"),
                      sliderInput("scale1", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit1", "Run Item Analysis", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,
                      # tableOutput("summary_table")
               )
             )
    ),
    tabPanel("Paired Samples Data Analysis",
             fluidRow(
               column(12, br(),
                      fileInput("file2", "Choose Pre-test Data File"),
                      fileInput("file3", "Choose Post-test Data File"),
                      sliderInput("scale2", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit2", "Run Paired Samples Data Analysis", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,

               )
             )
    ),
    tabPanel("Independent Samples Data Analysis",
             fluidRow(
               column(12, br(),
                      fileInput("file4", "Choose Treatment-group Data File"),
                      fileInput("file5", "Choose Control-group Data File"),
                      sliderInput("scale3", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit3", "Run Independent Samples Data Analysis", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,

               )
             )
    ),
    tabPanel("One-way ANCOVA",
             fluidRow(
               column(12, br(),
                      fileInput("file6", "Choose Treatment-group Pre-test Data File"),
                      fileInput("file7", "Choose Treatment-group Post-test Data File"),
                      fileInput("file8", "Choose Control-group Pre-test Data File"),
                      fileInput("file9", "Choose Control-group Post-test Data File"),
                      sliderInput("scale4", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit4", "Run One-way ANCOVA", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,

               )
             )
    ),
    tabPanel("One-way Repeated Measures ANOVA",
             fluidRow(
               column(12, br(),
                      fileInput("file10", "Choose Pre-test Data File"),
                      fileInput("file11", "Choose Post-test Data File"),
                      fileInput("file12", "Choose Post2-test Data File"),
                      sliderInput("scale5", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit5", "Run One-way Repeated Measures ANOVA", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,

               )
             )
    ),
    tabPanel("Analysis of Demographic Group Differences",
             fluidRow(
               column(12, br(),
                      fileInput("file13", "Choose Assessment Data File"),
                      fileInput("file14", "Choose Demographic Data File"),
                      sliderInput("scale6", "Cutoff Value", min = 0, max = 1, value = 0.15),
                      selectInput("column", "Group Variable Name", choices = NULL),
                      # textInput("text7", "Group Variable Name"),
                      h3("___________________________", style = "color: blue;"),
                      actionButton("submit6", "Run Analysis of Demographic Group Differences", style = "font-size: 14px;", style = "font-weight: bold;", style = "color: white;", style = "background-color: blue;"),
                      height = c(-100,0,0,0),
                      spacing = c(-100,0,0,0)
               ),
               column(12,

               )
             )
    )
  )
)

server <- function(input, output, session) {
  data1 <- reactive({
    input$file1$datapath
  })
  # data11 <- reactive({
  #   inFile <- input$file14
  #   if (is.null(inFile)) return(NULL)
  #   read.csv(inFile$datapath)
  # })
  data2 <- reactive({
    input$file2$datapath
  })
  data3 <- reactive({
    input$file3$datapath
  })
  data4 <- reactive({
    input$file4$datapath
  })
  data5 <- reactive({
    input$file5$datapath
  })
  data6 <- reactive({
    input$file6$datapath
  })
  data7 <- reactive({
    input$file7$datapath
  })
  data8 <- reactive({
    input$file8$datapath
  })
  data9 <- reactive({
    input$file9$datapath
  })
  data10 <- reactive({
    input$file10$datapath
  })
  data11 <- reactive({
    input$file11$datapath
  })
  data12 <- reactive({
    input$file12$datapath
  })
  data13 <- reactive({
    input$file13$datapath
  })
  data14 <- reactive({
    input$file14$datapath
  })
  data <- reactive({
    inFile <- input$file14
    if (is.null(inFile)) return(NULL)
    read.csv(inFile$datapath)
  })


  scale1 <- reactive({
    input$scale1
  })
  scale2 <- reactive({
    input$scale2
  })
  scale3 <- reactive({
    input$scale3
  })
  scale4 <- reactive({
    input$scale4
  })
  scale5 <- reactive({
    input$scale5
  })
  scale6 <- reactive({
    input$scale6
  })
  # text7 <- reactive({
  #   input$text7
  # })

  ## Item Analysis

  # output$summary_table <- renderTable({
  #   summary(data11())
  # })


  observeEvent(input$submit1, {
    if (is.null(data1())) {
      return(NULL)
    }
    # data1 <- file1()
    item_analysis(score_csv_data = data1(),
                  m_cutoff = scale1()
                  )
  })

  ## Paired Samples Data Analysis

  observeEvent(input$submit2, {

    if (is.null(data2())) {
      return(NULL)
    }
    if (is.null(data3())) {
      return(NULL)
    }

    paired_samples(pre_csv_data = data2(),
                   post_csv_data = data3(),
                   m_cutoff = scale2()
                   )
  })

  ## Independent Samples Data Analysis

  observeEvent(input$submit3, {

    if (is.null(data4())) {
      return(NULL)
    }
    if (is.null(data5())) {
      return(NULL)
    }

    independent_samples(treat_csv_data = data4(),
                        ctrl_csv_data = data5(),
                        m_cutoff = scale3()
                        )
  })

  ## One-way ANCOVA

  observeEvent(input$submit4, {

    if (is.null(data6())) {
      return(NULL)
    }
    if (is.null(data7())) {
      return(NULL)
    }
    if (is.null(data8())) {
      return(NULL)
    }
    if (is.null(data9())) {
      return(NULL)
    }

    one_way_ancova(treat_pre_csv_data = data6(),
                   treat_post_csv_data = data7(),
                   ctrl_pre_csv_data = data8(),
                   ctrl_post_csv_data = data9(),
                   m_cutoff = scale4()
                   )
  })

  ## One-way Repeated Measures ANOVA

  observeEvent(input$submit5, {

    if (is.null(data10())) {
      return(NULL)
    }
    if (is.null(data11())) {
      return(NULL)
    }
    if (is.null(data12())) {
      return(NULL)
    }

    one_way_repeated_anova(treat_pre_csv_data = data10(),
                           treat_post_csv_data = data11(),
                           treat_post2_csv_data = data12(),
                           m_cutoff = scale5()
                           )
  })

  ## Demographic Group Differences

  observe({
    column_names <- names(data())[-1]
    updateSelectInput(session, "column", choices = column_names)
  })

  observeEvent(input$submit6, {

    if (is.null(data13())) {
      return(NULL)
    }
    if (is.null(data14())) {
      return(NULL)
    }

    demo_group_diff(score_csv_data = data13(),
                    group_csv_data = data14(),
                    m_cutoff = scale6(),
                    group_name = input$column
                    )
  })

}

shinyApp(ui, server)
